### PR TITLE
Move to PEP-621 and pyproject and cfg files

### PR DESCRIPTION
**Pull request by kinow**
Ref: #6

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](http://claassistant.ecmwf.int:80/ecmwf/pyflow?pullRequest=7) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://claassistant.ecmwf.int:80/ecmwf/pyflow?pullRequest=7) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://claassistant.ecmwf.int:80/check/ecmwf/pyflow?pullRequest=7) it.</sub>

File: setup.cfg [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    requests
+
+[options.packages.find]
+include = pyflow*
+
+[options.extras_require]
+diagrams =
+    graphviz
+tests =
+    pytest
```
Comment (kinow) [inline setup.cfg:27]: We don't have `pytest` in the `setup.py`, but I think it should be fine to add it here.

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+long_description = file: README.md
+long_description_content_type=text/markdown
+url = https://pyflow-workflow-generator.readthedocs.io/en/
+
+[options]
+packages = find_namespace:
+include_package_data = True
+install_requires =
+    jinja2
+    requests
```
Comment (kinow) [inline setup.cfg:18]: Maybe we could also pin the versions of the dependencies? i.e. `jinja2==2.*` or so

File: setup.cfg [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+version = attr: pyflow/version.__version__
+author = European Centre for Medium-Range Weather Forecasts (ECMWF)
+author_email = software.support@ecmwf.int
+license = Apache 2.0
+license_files = LICENSE
+description = Create pythonic ecFlow suites
+long_description = file: README.md
+long_description_content_type=text/markdown
+url = https://pyflow-workflow-generator.readthedocs.io/en/
+
```
Comment (kinow) [inline setup.cfg:12]: Missing platform, classifiers, and other project information that appears in the PyPI project page. Maybe another improvement for this change.

File: setup.cfg [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+include = pyflow*
+
+[options.extras_require]
+diagrams =
+    graphviz
+tests =
+    pytest
+all =
+    %(diagrams)s
+    %(tests)s
```
Comment (kinow) [inline setup.cfg:30]: For convenience, so that devs/contributors can just `pip install -e .[all]` and get all dependencies installed.

**Comment by kinow** [comment]
>changed the base branch from master to develop 

Argh. Thanks!

**Review by kinow — COMMENTED** [review]
I actually started this branch to add [`mypy.ini`](#5) and scan the project, but digressed, went to the ecFlow repo to [ask about publishing it to PyPI](https://github.com/ecmwf/ecflow/issues/24), to the ecFlow Conda forge to [remind that the doc URL needs to be updated](https://github.com/conda-forge/ecflow-feedstock/pull/39), and eventually came back here to create this draft PR :slightly_smiling_face: :earth_asia: :airplane: 

Feel free to close it without merging, or using/adapting anything from this PR.

Thanks!

**Comment by kinow** [comment]
p.s.: to test this PR, in case you are using an older version of `pip`, you may have to `pip install -U pip` before trying to install pyFlow from this branch.

File: setup.cfg (hunk 2 above)
Comment (kinow) [reply setup.cfg:18]: And I was going to add `ecflow` from PyPI when I realized it's not deployed there. I went to ecmwf/ecflow to try to install it from a tag, when I remembered it's a bind to the C++ code. I guess it should be possible to upload a version for that too in PyPI?

File: setup.cfg [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+author_email = software.support@ecmwf.int
+license = Apache 2.0
+license_files = LICENSE
+description = Create pythonic ecFlow suites
+long_description = file: README.md
+long_description_content_type=text/markdown
+url = https://pyflow-workflow-generator.readthedocs.io/en/
+
+[options]
+packages = find_namespace:
```
Comment (oiffrig) [inline setup.cfg:14]: Any particular reason why find_namespace instead of find? Not sure it matters much though.

File: setup.cfg [hunk 6]
```diff
@@ -0,0 +1,30 @@
+[metadata]
+name = pyflow
+version = attr: pyflow/version.__version__
```
Comment (oiffrig) [inline setup.cfg:3]: Shouldn't that be `attr: pyflow.version.__version__`?

File: setup.py [hunk 7]
```diff
@@ -1,31 +0,0 @@
...
-
-def read(fname):
-    file_path = os.path.join(os.path.dirname(__file__), fname)
-    return io.open(file_path, encoding="utf-8").read()
-
-
-exec(open("pyflow/version.py").read())
-
-
-setup(
```
Comment (oiffrig) [inline setup.py:15]: I would keep a trivial `setup.py` for compatibility with older versions of `pip` when using editable installs. It should be enough to have
```python
from setuptools import setup
setup()
```

File: setup.cfg (hunk 5 above)
Comment (kinow) [reply setup.cfg:14]: Not really, sorry. I used another project that uses Python namespaces (PEP 382). Should work with `find` as I believe pyflow doesn't use namespaces.

File: setup.cfg (hunk 6 above)
Comment (kinow) [reply setup.cfg:3]: Ops, you are right, sorry!

File: setup.py (hunk 7 above)
Comment (kinow) [reply setup.py:15]: Sounds good to me. Will update this PR with your feedback @oiffrig , thanks!!

File: setup.cfg (hunk 5 above)
Comment (kinow) [reply setup.cfg]: Done.

File: setup.cfg (hunk 6 above)
Comment (kinow) [reply setup.cfg]: Done.

File: setup.py (hunk 7 above)
Comment (kinow) [reply setup.py:15]: Done.

**Review by kinow — COMMENTED** [review]
Updated code as per @oiffrig 's comments, rebased on `develop` and squashed. Created a new venv, did a `pip install -e .` and it was installed successfully. Tried running the tests but that failed due to missing deps, I think (e.g. ecflow). Hopefully CI will catch any errors.

**Pull request merged**